scripts/buildorder.py

```diff
@@ -262,104 +278,132 @@ def read_packages_from_directories(directories, fast_build_mode, full_buildmode)
                 dep_pkg.needed_by.add(pkg)
     return pkgs_map
 
-def generate_full_buildorder(pkgs_map):
-    "Generate a build order for building all packages."
-    build_order = []
-
-    # List of all TermuxPackages without dependencies
-    leaf_pkgs = [pkg for pkg in pkgs_map.values() if not pkg.deps]
-
-    if not leaf_pkgs:
-        die('No package without dependencies - where to start?')
-
-    # Sort alphabetically:
-    pkg_queue = sorted(leaf_pkgs, key=lambda p: p.name)
-
-    # Topological sorting
-    visited = set()
-
-    # Tracks non-visited deps for each package
-    remaining_deps = {}
-    for name, pkg in pkgs_map.items():
-        remaining_deps[name] = set(pkg.deps)
-        for subpkg in pkg.subpkgs:
-            remaining_deps[subpkg.name] = set(subpkg.deps)
-
-    while pkg_queue:
-        pkg = pkg_queue.pop(0)
-        if pkg.name in visited:
-            continue
-
-        # print("Processing {}:".format(pkg.name), pkg.needed_by)
-        visited.add(pkg.name)
-        build_order.append(pkg)
-
-        for other_pkg in sorted(pkg.needed_by, key=lambda p: p.name):
-            # Remove this pkg from deps
-            remaining_deps[other_pkg.name].discard(pkg.name)
-            # ... and all its subpackages
-            remaining_deps[other_pkg.name].difference_update(
-                [subpkg.name for subpkg in pkg.subpkgs]
-            )
-
-            if not remaining_deps[other_pkg.name]:  # all deps were already appended?
-                pkg_queue.append(other_pkg)  # should be processed
-
-    if set(pkgs_map.values()) != set(build_order):
-        print("ERROR: Cycle exists. Remaining: ", file=sys.stderr)
-        for name, pkg in pkgs_map.items():
-            if pkg not in build_order:
-                print(name, remaining_deps[name], file=sys.stderr)
-
-        # Print cycles so we have some idea where to start fixing this.
-        def find_cycles(deps, pkg, path):
-            """Yield every dependency path containing a cycle."""
-            if pkg in path:
-                yield path + [pkg]
-            else:
-                for dep in deps[pkg]:
-                    yield from find_cycles(deps, dep, path + [pkg])
+def remove_only_installing_deps(pkgs_map, deps):
+    """Completely removing dependencies that can only be
+    installed during package compilation, and setting their
+    dependencies from the `pkg_deps` value."""
 
-        cycles = set()
-        for pkg in remaining_deps:
-            for path_with_cycle in find_cycles(remaining_deps, pkg, []):
-                # Cut the path down to just the cycle.
-                cycle_start = path_with_cycle.index(path_with_cycle[-1])
-                cycles.add(tuple(path_with_cycle[cycle_start:]))
-        for cycle in sorted(cycles):
-            print(f"cycle: {' -> '.join(cycle)}", file=sys.stderr)
+    pkgs_only_installing = set()
```

Comment:
The `remove_only_installing_deps` function - it removes dependencies that can only be installed, and at the same time it adds dependencies (from the `pkg_deps` value) of those removed dependencies.

> But what is pkgs_only_installing, and when does it come into play? What goes into this set, and under what conditions?

The `pkgs_only_installing` variable stores packages, that can only be installed, to avoid an infinite processing loop.

Reply:
I changed the description of the `remove_only_installing_deps` function and added a comment to this variable